sdk/storage/azure-storage-blob-nio/src/main/java/com/azure/storage/blob/nio/AzureFileSystemProvider.java

```diff
@@ -325,13 +369,184 @@ public void delete(Path path) throws IOException {
+                + "can be just a root directory. Source: %s. Destination: %s.", source.toString(),
+                destination.toString())));
+        }
+
+        // Build clients.
+        BlobClient sourceBlob = ((AzurePath) source).toBlobClient();
+        BlobClient destinationBlob = ((AzurePath) destination).toBlobClient();
+
+        // Check destination is not a directory with children.
+        DirectoryStatus destinationStatus = checkDirStatus(destinationBlob);
+        if (destinationStatus.equals(DirectoryStatus.NOT_EMPTY)) {
+            throw Utility.logError(logger, new DirectoryNotEmptyException(destination.toString()));
+        }
+
+        /*
+        Set request conditions if we should not overwrite. We can error out here if we know something already exists,
+        but we will also create request conditions as a safeguard against overwriting something that was created
+        between our check and put.
+         */
+        BlobRequestConditions requestConditions = null;
+        if (!replaceExisting) {
+            if (!destinationStatus.equals(DirectoryStatus.DOES_NOT_EXIST)) {
+                throw Utility.logError(logger, new FileAlreadyExistsException(destination.toString()));
+            }
+            requestConditions = new BlobRequestConditions().setIfNoneMatch("*");
+        }
+
+        /*
+        More path validation
+
+        Check that the parent for the destination exists. We only need to perform this check if there is nothing
+        currently at the destination, for if the destination exists, its parent at least weakly exists and we
+        can skip a service call.
+         */
+        if (destinationStatus.equals(DirectoryStatus.DOES_NOT_EXIST) && !checkParentDirectoryExists(destination)) {
+            throw Utility.logError(logger, new IOException("Parent directory of destination location does not exist."
+                + "The destination path is therefore invalid. Destination: " + destination.toString()));
+        }
+
+        /*
+        Try to copy the resource at the source path.
+
+        There is an optimization here where we try to do the copy first and only check for a virtual directory if
+        there's a 404. In the cases of files and concrete directories, this only requires one request. For virtual
+        directories, however, this requires three requests: failed copy, check status, create directory. Depending on
+        customer scenarios and how many virtual directories they copy, it could be better to check the directory status
+        first and then do a copy or createDir, which would always be two requests for all resource types.
+         */
+        try {
+            SyncPoller<BlobCopyInfo, Void> pollResponse =
+                destinationBlob.beginCopy(sourceBlob.getBlobUrl(), null, null, null, null, requestConditions, null);
+            pollResponse.waitForCompletion(Duration.ofSeconds(30));
```

Comment:
Is there a reason for the duration being 30 here?

Reply:
Nope. Just a magic number plus being a little lazy about adding a constant/option. I figure we can add an option for it if/when people are dissatisfied with this. Or I can add an option now. :P

Reply:
might at least be helpful to make it a constant so we dont have to dig around for them later. ya know?